src/backend/distributed/executor/directed_acylic_graph_execution_logic.c

```diff
@@ -0,0 +1,228 @@
+ * is already executed, it is not added to the result.
+ */
+static List *
+FindExecutableTasks(List *allTasks, HTAB *completedTasks)
+{
+	List *curTasks = NIL;
+	ListCell *taskCell = NULL;
+	TaskHashKey taskKey;
+
+	foreach(taskCell, allTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		FillTaskKey(&taskKey, task);
+
+		if (IsAllDependencyCompleted(task, completedTasks) &&
+			!IsTaskAlreadyCompleted(task, completedTasks))
+		{
+			curTasks = lappend(curTasks, task);
+		}
+	}
+
+	return curTasks;
+}
+
+
+/*
+ * AddCompletedTasks adds the givens tasks to completedTasks HTAB.
+ */
+static void
+AddCompletedTasks(List *curCompletedTasks, HTAB *completedTasks)
+{
+	ListCell *taskCell = NULL;
+	TaskHashKey taskKey;
+	bool found;
+
+	foreach(taskCell, curCompletedTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		FillTaskKey(&taskKey, task);
+		hash_search(completedTasks, &taskKey, HASH_ENTER, &found);
+	}
+}
+
+
+/*
+ * CreateTaskHashTable creates a HTAB with the necessary initialization.
+ */
+static HTAB *
+CreateTaskHashTable()
+{
+	uint32 hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT | HASH_COMPARE);
+	HASHCTL info = InitHashTableInfo();
+	return hash_create("citus task completed list (jobId, taskId)",
+					   64, &info, hashFlags);
+}
+
+
+/*
+ * IsTaskAlreadyCompleted returns true if the given task
+ * is found in the completedTasks HTAB.
+ */
+static bool
+IsTaskAlreadyCompleted(Task *task, HTAB *completedTasks)
+{
+	TaskHashKey taskKey;
+	bool found;
+
+	FillTaskKey(&taskKey, task);
+	hash_search(completedTasks, &taskKey, HASH_ENTER, &found);
+	return found;
+}
+
+
+/*
+ * IsAllDependencyCompleted return true if the given task's
+ * dependencies are completed.
+ */
+static bool
+IsAllDependencyCompleted(Task *targetTask, HTAB *completedTasks)
+{
+	ListCell *taskCell = NULL;
+	bool found = false;
+	TaskHashKey taskKey;
+
+	foreach(taskCell, targetTask->dependedTaskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		FillTaskKey(&taskKey, task);
+
+		hash_search(completedTasks, &taskKey, HASH_FIND, &found);
+		if (!found)
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+
+static void
+FillTaskKey(TaskHashKey *taskKey, Task *task)
+{
+	taskKey->jobId = task->jobId;
+	taskKey->taskId = task->taskId;
+}
+
+
+/*
+ * InitHashTableInfo returns hash table info, the hash table is
+ * configured to be created in the CurrentMemoryContext so that
+ * it will be cleaned when this memory context gets freed/reset.
+ */
+static HASHCTL
+InitHashTableInfo()
+{
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(TaskHashKey);
+	info.entrysize = sizeof(TaskHashEntry);
+	info.hash = TaskHash;
+	info.match = TaskHashCompare;
+	info.hcxt = CurrentMemoryContext;
+
+	return info;
+}
+
+
+static uint32
+TaskHash(const void *key, Size keysize)
+{
+	TaskHashKey *taskKey = (TaskHashKey *) key;
+	uint32 hash = 0;
+
+	hash = hash_combine(hash, hash_uint32((uint32) taskKey->jobId));
```

Comment:
Should we try to preserve the 64-bitness here?

Reply:
hmm at first I couldnt find a method like hash_uint64 but I guess we do it this way :

```
hash = hash_combine(hash, hash_any((unsigned char *) &taskKey->jobId,
									   sizeof(int64)));
```